### PR TITLE
UKMO Staging May 2021

### LINKED ISSUE #361: Extra restart - checks
When reading a restart file containing extra fields, such as an OASIS coupled configuration that sends fields to other models via coupling at t+0 does by default, there is a risk that the fields that are read do not correspond to the right field. This happens when the restart file contains a different set of extra fields than the model is trying to read.

At the moment there is no check for this possibility, and the result is that a coupled run can silently be using the wrong initial fields, which can easily cause a crash. The purpose of this issue is to, when reading a extra fields in a restart file, give a warning if the field is supposed to be read but it is not present, skipping it, and an error if in this same situation the configuration requires this field to be present in the restart because it is a coupling field that is sent to other models at T+0. In principle a coupling field can be initialized to a default value rather than to sensible values obtained by a previous model run, but this often leads to instabilities of the coupled model resulting in a crash. 

This can easily be achieved by writing in the header of the restart file the FLOGRR vector, and comparing this variable with the current value of the variable when reading the restart file.

### LINKED ISSUE #377: Wrong switches for coupled field descriptors
**Describe the bug**
The routine W3FLDTOIJ returns a set of I,J indexes identifying the chain of characters that describe a particular output/coupled/extra restart field. In the case of the wave height and Charnock coefficient, identified by the descriptors 'HS' and 'CHA', there are extra descriptors for these variables when running in coupled mode, which add an extra 'A' or 'O' at the start of the descriptor to indicate if they are used in ocean or atmosphere coupling. These extra descriptors are protected by the compilation switches OASACM and OASOCM. In the case of the 'OHS' and 'OCHA' descriptors the switches on the line that set the value of the J index is OASACM, which is wrong and it should be OASOCM.

**To Reproduce**
Run any ocean/wave or atmosphere/wave coupled configuration (not an atmosphere/ocean/wave fully coupled configuration) coupling or sending to the output the wave height or/and the Charnock coefficient.

**Expected behavior**
When compiling in debug mode, there should be a warning saying something like (e.g. the GNU compiler):

  Subscript -1 is out of range for dimension 2 for array
  'flg2d(,*)' at line 619 in file 'w3iogomd.F90' with bounds 1:20.

This happens because the J index in the aforementioned routine is wrongly set to -1 for the wave height and the Charnock coefficient. This means that this field will be ignored in the output, and will not be identified as a field that is coupled, resulting in error messages from the OASIS coupler. Furthermore, this might lead to writing data in a wrong memory address and resulting in potential crashes or unexpected results.

**Screenshots**
Non applicable

**Additional context**
No additional context

### LINKED ISSUE #378: ACTION array allocation too small
**Describe the bug**
The `ACTION` array in the `wminitmd.f90` subroutine `WMINITNML` is allocated with an upper bound of 9:
https://github.com/NOAA-EMC/WW3/blob/482c500da51c2caf41ce3f8f011855bb4ab38d76/model/ftn/wminitmd.ftn#L4081

but it is later accessed with array indices > 9 e.g.:
https://github.com/NOAA-EMC/WW3/blob/482c500da51c2caf41ce3f8f011855bb4ab38d76/model/ftn/wminitmd.ftn#L4167

The `WMINIT` subroutine also contains an ACTION array, however this is defined as a static array of the correct size 11:
https://github.com/NOAA-EMC/WW3/blob/482c500da51c2caf41ce3f8f011855bb4ab38d76/model/ftn/wminitmd.ftn#L429

**To Reproduce**
Running a multi-grid model using NAMELIST input (ww3_multi.nml) causes a runtime error.
The error does **not** occur when using `ww3_multi.inp` (WMINT is called rather than WMINITNML)

**Expected behavior**
Not segfault when using ww3_multi.nml

### LINKED ISSUE #383: Ice and other model coupling - wrong warning
**Describe the bug**
When coupling to an ice and other model, the program will generate a warning message for each variable XXX sent to models other than ice and for each processor even if the run parameters are right, which can be quite confusing for inexperienced users:

 *** WAVEWATCH III WARNING IN W3IGCMMD : 
     REQUESTED COUPLING SENT FIELD XXX WAS NOT RECOGNIZED.

This happens because the routine identifying sent ice coupling fields gives this warning if any coupling field is not an ice coupled field. A similar line is present in the equivalent receive ice fields routine, which has been correctly commented.

Although this is not a proper bug in the sense that the code will not generate incorrect results, it is wrong because a warning is printed when it should not, which can be confusing - OASIS will provide proper error messages in case the coupling configuration is problematic.

**To Reproduce**
Run any configuration coupled to ice and other model (atmosphere or ocean).

**Expected behavior**
With this fix the warning will be removed, and the run will continue successfully - incorrect coupling configurations will be detected by OASIS.

**Screenshots**
N/A

**Additional context**
N/A

### LINKED ISSUE #390: OASIS send condition
**Describe the bug**
When running a coupled configuration, there is a series of conditions that have to be fulfilled before calling the oasis_send routine. One of this conditions is problematic, as when the run length is three days or more this condition is not fulfilled when it should. Such condition is:

MOD(ID_OASIS_TIME/DTOUT(7),1.0)

**To Reproduce**
Any coupled configuration sending information from the wave to other model via coupling running for three days or longer.

**Expected behavior**
The run crashes with an OASIS error after more than two days saying that the order of coupling is not fulfilled, as one call to oasis_send has been skipped.

**Screenshots**
N/A

**Additional context**
N/A

### LINKED ISSUE #397: floating overflow on regtest ww3_tp2.8
**Describe the bug**
a floating overflow error appears when running the regtest ww3_tp2.8.
this is due to the call of the subroutine INTERP in ww3_prnc

**To Reproduce**
./bin/run_test -o netcdf -N -c datarmor_mpt_debug ../model ww3_tp2.8

**Expected behavior**
I suspect the muliplication of 2*fillvalue*fillvalue produce a too large value

**Screenshots**
   Screen output routed to  regtests/ww3_tp2.8/work/ww3_prnc.out
forrtl: error (72): floating overflow
Image              PC                Routine            Line        Source             
ww3_prnc           00000000009B8D9E  Unknown               Unknown  Unknown
libpthread-2.19.s  00007FFFEB3E6870  Unknown               Unknown  Unknown
ww3_prnc           0000000000484D28  MAIN__                   1648  ww3_prnc.F90
ww3_prnc           000000000040BF1E  Unknown               Unknown  Unknown
libc-2.19.so       00007FFFEB050B25  __libc_start_main     Unknown  Unknown
ww3_prnc           000000000040BE29  Unknown               Unknown  Unknown
./bin/run_test: line 1253:  6951 Aborted                 $runprog $path_e/$prog > $ofile

line 1648 in ww3_prnc.F90 if : 
                CALL INTERP(MXM, MYM, XC**2+YC**2, IX21, IX22,         &
                               IY21, IY22, RD11, RD12, RD21, RD22,     &
                               2.0*FILLVALUE*FILLVALUE, A3)

### LINKED ISSUE #398: Clarify NRQMAX calculation
The calculation of NRQMAX in `w3initmd` is a quite confusing and not at all clear:

https://github.com/NOAA-EMC/WW3/blob/3b949d36a00f5e0cf6a7b79f382c491b8b2e1800/model/ftn/w3initmd.ftn#L2056-L2061

This line should:
  a) Be reformatted over multiple lines, each line representing a single output group.
  b) have an explanatory comment added.

The explanation comment should also make it clear why most of the fields in group 1 do not contribute to the total NRQMAX count (as they are input variables already available to all processors).

### LINKED ISSUE #420: Out-of-bounds array access in ww3_prnc
**Describe the bug**
There is an out-of-bounds array access occurring in the `XC` variable of ww3_prnc in the INTERP subroutine.
In situations where the interpolation weights are all zero for a point, the wrong indices are being used to access the XC array.

This is likely the reason why some regtests, such as ww3_tp2.15, are showing non b4b results in the generated wind.ww3 files.

**To Reproduce**
Run `ww3_tp2.15/5km` with runtime array bound checking.

**Expected behavior**
The GNU compiler on the UKMO HPC shows the following runtime error:
```
At line 2106 of file ww3_prnc.F90
Fortran runtime error: Index '122' of dimension 2 of array 'xc' above upper bound of 121
```

Whilst this does not cause a segfault in normal operation, the produced wind.ww3 file will have values in it from undefined regions of memory.

### LINKED ISSUE #430: [ww3_ounf] RTD pole information being generated for non-rotated pole grids
**Describe the bug**
If a model is compiled with the `RTD` switch but is formulated on a standard pole* the `ww3_ounf` program incorrectly generates extra `standard_latitude` and `standard_longitude` variables.

_* e.g. for generation of rotated pole boundary conditions from standard pole model_

**To Reproduce**
Steps to reproduce the behaviour:
Run a model on a standard grid, but compile with the `RTD` switch

**Expected behaviour**
The `standard_latitude` and `standard_longitude` variables should not be genereted.
Additionally, the following attributes for `latitude` and `longitude` are incorrect:
```
longitude:long_name = "longitude in rotated pole grid" ;
longitude:standard_name = "grid_longitude" ;
```

They should be simply (similarly for latitude):
```
longitude:long_name = "longitude" ;
longitude:standard_name = "longitude" ;
```

### PR DESCRIPTION
**Pull request by ukmo-ccbunney**
# Pull Request Summary
Bug fixes and new features from UKMO fork comprising the following PRs:
   - :bug: ukmo-waves/WW3#32 : Erroneous rotated pole netCDF variables on standard pole grid [ww3_ounf]
   - 🐛 ukmo-waves/WW3#31 : WW3_PRNC Interp bug
   - 🐛 ukmo-waves/WW3#29 : Fix float overflow in interp call in w3_prnc
   - 🐛 ukmo-waves/WW3#28 : Clarification of NRQMAX calculation
   - 🐛 ukmo-waves/WW3#26 : Update OASIS coupling condition to works correctly for longer runtimes
   - ⭐ ukmo-waves/WW3#25 : Two new features in WW3_OUNF
   - 🐛 ukmo-waves/WW3#24 : Remove confusing warning when ice coupling 
   - 🐛 ukmo-waves/WW3#23 : Fix incorrect allocation of ACTION array
   - 🐛 ukmo-waves/WW3#22 : Fix to oasis compilation switches.
   - ⭐ ukmo-waves/WW3#21 : Add support for Intel compiler on Cray XC
   - ⭐ ukmo-waves/WW3#20 : Support for Application Level Placement Scheduler (ALPS)

  ## Description
See individual PR links for details of specific changes.

### Issue(s) addressed
- fixes NOAA-EMC#430
- fixes NOAA-EMC#420
- fixes NOAA-EMC#397
- fixes NOAA-EMC#398
- fixes NOAA-EMC#361
- fixes NOAA-EMC#390
- fixes NOAA-EMC#383
- fixes NOAA-EMC#378
- fixes NOAA-EMC#377

### Check list  
* Is your feature branch up to date with the authoritative repository (NOAA/develop)?
* Make sure you have checked the [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop) and [updating version number](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-updating-version-number)
* Reviewers: @mentions of suggested reviewers of the proposed changes

### Testing
* How were these changes tested?
  *  **Targetted regression tests and full regtest matrix**
* Are the changes covered by regression tests? (If not, why? Do new tests need to be added?)
  *  **Yes**
* If a new feature was added, was a new regression test added?
  *  **N/A**
* Have regression tests been run?
  *  **Yes**
* Which compiler / HPC you used to run the regression tests in the PR?
  *  **Cray XC; GNU compiler**
   

### Regression tests
Full regression test matrix uploaded here: 
[matrixComp_ukmo_staging_may2021.zip](https://github.com/NOAA-EMC/WW3/files/6678122/matrixComp_ukmo_staging_may2021.zip)

Known non B4B regtests:
  - **mww3_test_03**

Expected differences:
  - **ww3_tnc1** - Modified to test removal of MAPSTA variable and optional fill value (ukmo-waves/WW3#25)
  - **ww3_tp2.10**, **ww3_tp2.14** and **ww3_tp2.16**: Output of string `MAPSTA output disabled for SMC grids` in `ww3_ounf.out`.

### CODE REVIEW

**Comment by aliabdolali** [comment]
@ukmo-ccbunney I ran the tests and here are the tests that I need your comments on them:

[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/6683662/matrixDiff.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/6683663/matrixCompSummary.txt)
[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/6683664/matrixCompFull.txt)

---------------------------------------------------

`ww3_tnc1/./work `


---------------------------------------------------

`ww3_tp2.10/./work_MPI                     (1 files differ)`
     _ww3_ounf.out differ._
`ww3_tp2.14/./work_OASACM6                     (1 files differ)`
     _ww3_ounf.out differ._ 

the diff is _MAPSTA output disabled for SMC grids_

---------------------------------------------------

```
ww3_tp2.15/./work_PR3_UQ_MPI                     (1 files differ)
ww3_tp2.15/./work_PR3_UQ_RHO_MPI                     (1 files differ)
ww3_tp2.15/./work_PR3_UQ_RHO                     (1 files differ)
ww3_tp2.15/./work_PR3_UQ                     (1 files differ)
```
_wind.ww3_ is different and I think this is the same problem with what we had just on our HPC. It is interesting to see we just had this problem for `ww3_tp2.15/./work_5km` and `ww3_tp2.15/./work_MPI_5km` and now it is expanded to more `ww3_tp2.15`s

---------------------------------------------------

`ww3_tp2.16/./work_MPI_OMPH                     (1 files differ)`
_ww3_ounf.out_
the diff is _MAPSTA output disabled for SMC grids_
`ww3_tp2.16/./work_MPI                     (1 files differ)`
_ww3_ounf.out_
the diff is _MAPSTA output disabled for SMC grids_

---------------------------------------------------

the know non b4b tests are:

```
mww3_test_07
mww3_test_03
ww3_tp2.10
ww3_tp2.15
```

**Comment by ukmo-ccbunney** [comment]
Hi @aliabdolali 
Your regression test results look consistent with what I expected, except for the `wind.ww3` differences in ww3_tp2.15.
I don't think I have seen these differences before, but I am rerunning some regtests to see if I can reproduce.

It must be an issue with `ww3_prnc`, or perhaps something inconsistent in the `mod_def.ww3` file generated by `ww3_grid`?

**Comment by ukmo-ccbunney** [comment]
@aliabdolali  - OK - I think I know why there are differences in the wind.ww3 program: I've spotted a bug in ww3_prnc that is causing an array out-of-bounds access. I'm just running some tests and I'll hopefully have a fix today.

See: NOAA-EMC/WW3#420

File: model/bin/cmplr.env [hunk 1]
```diff
@@ -189,11 +174,19 @@ if [ "$cmplr" == "intel" ]          || [ "$cmplr" == "intel_debug" ]    || [ "$c
   if [ ! -z "$(echo $cmplr | grep datarmor)" ] ; then
     optc="$optc -xcore-avx2"
     optl="$optl -xcore-avx2"
+  elif [ ! -z "$(echo $cmplr | grep wcoss_cray)" ] ; then
```
Comment (JessicaMeixner-NOAA) [inline model/bin/cmplr.env:177]: we should probably do some sanity checks by running ufs-weather-model on wcoss @aliabdolali

**Review by JessicaMeixner-NOAA — COMMENTED** [review]
I think we need to do a sanity check on wcoss based on the changes in cmplr.env although they all look fine.  Thank you so much for the added NRQMAX explanation --- that's super helpful!

File: model/bin/cmplr.env (hunk 1 above)
Comment (ukmo-ccbunney) [reply model/bin/cmplr.env:177]: Probably a sensible thing to do!
A good first check is to make sure the list of "opts" in your generated `comp` and `link` scripts are the same...

**Comment by ukmo-ccbunney** [comment]
Updated to include ukmo-waves/WW3#31
I am re-running the regression tests.

**Comment by ukmo-ccbunney** [comment]
Regression tests have been re-run to include bugfix to ww3_prnc(  ukmo-waves#31)

Results are the same as previous, with the addition of changes to `ww3_tp2.15`; the `wind.ww3` file is slightly different (due to bugfix) and this also slightly affects the model outputs.

```
ww3_tp2.15/./work_PR3_UQ_RHO                     (7 files differ)
ww3_tp2.15/./work_PR3_UQ_RHO_MPI                     (5 files differ)
ww3_tp2.15/./work_PR3_UQ                     (6 files differ)
ww3_tp2.15/./work_MPI_5km                     (3 files differ)
ww3_tp2.15/./work_5km                     (5 files differ)
ww3_tp2.15/./work_PR3_UQ_MPI                     (38 files differ)
```

A quick sanity check on the outputs (work_5km; ww3.201407.nc) shows only minimal differences in the u-wind field (develop on left, feature branch on right).

image

Full regression test results here: 
[matrix_comp_staging_may2021_v2.zip](https://github.com/NOAA-EMC/WW3/files/6696191/matrix_comp_staging_may2021_v2.zip)

**Comment by aliabdolali** [comment]
@ukmo-ccbunney thanks, I am going to rerun them on our hpc and I'll check them within our coupling framework as a sanity check, hopefully today or tomorrow.
Your fix to ww3_tp2.15 will remove it from the list of know non b4b tests. Thanks for doing this.

**Comment by aliabdolali** [comment]
@ukmo-ccbunney I am going to check the ukmet/staging now. Please merge noaa/develop and solve the conflict.

**Comment by ukmo-ccbunney** [comment]
> 
> 
> @ukmo-ccbunney I am going to check the ukmet/staging now. Please merge noaa/develop and solve the conflict.

OK - all done.

**Comment by aliabdolali** [comment]
@ukmo-ccbunney 
I have segfault for the following with intel
run_test -b slurm -c hera.intel -S -T -o netcdf -i input_rho -w work_PR3_UQ_RHO -r ww3_shel ../model11 ww3_tp2.15

**Comment by ukmo-ccbunney** [comment]
> 
> 
> @ukmo-ccbunney
> I have segfault for the following with intel
> run_test -b slurm -c hera.intel -S -T -o netcdf -i input_rho -w work_PR3_UQ_RHO -r ww3_shel ../model11 ww3_tp2.15

That's odd. It ran OK for you in the https://github.com/ukmo-waves/WW3/pull/31 didn't it?
Were you using Intel for that test too?

**Comment by aliabdolali** [comment]
> > @ukmo-ccbunney
> > I have segfault for the following with intel
> > run_test -b slurm -c hera.intel -S -T -o netcdf -i input_rho -w work_PR3_UQ_RHO -r ww3_shel ../model11 ww3_tp2.15
> 
> That's odd. It ran OK for you in the [ukmo-waves#31](https://github.com/ukmo-waves/WW3/pull/31) didn't it?
> Were you using Intel for that test too?

yes, it was also intel, I will dig more

**Comment by ukmo-ccbunney** [comment]
> 
> 
> > > @ukmo-ccbunney
> > > I have segfault for the following with intel
> > > run_test -b slurm -c hera.intel -S -T -o netcdf -i input_rho -w work_PR3_UQ_RHO -r ww3_shel ../model11 ww3_tp2.15
> > 
> > 
> > That's odd. It ran OK for you in the [ukmo-waves#31](https://github.com/ukmo-waves/WW3/pull/31) didn't it?
> > Were you using Intel for that test too?
> 
> yes, it was also intel, I will dig more

@aliabdolali - I think I know what is happening here. The `PR3_UQ_RHO` variant of the `ww3_tp2.15` regtests reads in a "density.ww3" file generated by ww3_prnc.

Looking at the output from ww3_prnc, It looks like not all model grid points are covered by the input netCDF file, excerpt:
```
 *** WAVEWATCH-III WARNING W3PRNC :
     GRID POINT     39     2  18.84  40.25
 NOT COVERED BY INPUT GRID.


 *** WAVEWATCH-III WARNING W3PRNC :
     GRID POINT     20    27  15.65  43.76
 NOT COVERED BY INPUT GRID.


 *** WAVEWATCH-III WARNING W3PRNC :
     GRID POINT      2    30  12.28  44.20
 NOT COVERED BY INPUT GRID.
```
This means these points are probably getting set to `0.0` in the `density.ww3` file. A zero density is likely causing some problems with some calculations; I get a "floating point exception" error when running Intel debug. Setting a default of 0.0 is sensible for most fields, but not density!

I think a valid value was being set for these values prior to the https://github.com/ukmo-waves/WW3/pull/31 fix - purely by chance.

I am going to tag @ukmo-juan-castillo here as I am about to finish work and will be away until next week: Juan, the bugfix from the PR mentioned above is likely causing this issue. Are you able to have a look? One fix might be to sort out the input netCDF file so it actually covers all model grid points...?

Edit: Perhaps set all the missing cells in the density.nc file to 1.225?

**Comment by ukmo-juan-castillo** [comment]
Thank you for finding the problem, I will pass you the new file later and hopefully this will fix the problem.

**Comment by ukmo-juan-castillo** [comment]
I do not see any zeroes in the middle of the domain in the density file. This message must be related to the interpolation, I will need to investigate.

**Comment by aliabdolali** [comment]
for the difference ib _ww3_ufs1.1_ and _ww3_ufs1.1_ tests, I just plotted the wave outputs (wind, current, ice and Hs) and all have changed mostly close to the coastline. Did you change anything in the interpolation of forcing in ww3_prnc?

**Comment by ukmo-juan-castillo** [comment]
The interpolation along the coastlines will be influenced by issue #420 . If there are some forcing points with zero values, it must be because originally those points contained FillValues, as I do not see any points in the netcdf forcing file with zero values originally. This is probably a problem of a wrong land/sea mask of the input fields compared to what the model grid in the configuration expects.

If you send me a complete list of the points where the density is zero but should not (all of them should lie along coastlines; I would prefer indexes rather than lat-lon, but I can work with both) I can try filling those points with sensible values. I am afraid I can not reproduce your run here so I can not generate that list myself. Or if it is easier at least for testing, we can substitute all the fill values with a constant density and see if that works.

Thanks for finding the root of the problem.

**Comment by ukmo-juan-castillo** [comment]
I can reproduce now the segmentation fault and run some extra experiments. I agree that the best solution is modifying the input fields so that they have appropriate values in all wet points. It would be the responsibility of the user to make sure that this is the case in his configurations. I will upload the files when I have them ready.

**Comment by ukmo-ccbunney** [comment]
A future improvement might be to have a "default value" for interpolated model points that are all FILLVALUE in the source grid. Currently, this is set to zero in the INTERP routine, but we could set it to a value that is relevant to the input field instead (zero in most cases, but non-zero for fields like RHO).

We could set the default value in field checking block starting at line 420 (.nml) and 561 (.inp)

**Comment by ukmo-juan-castillo** [comment]
After generating a new density file there are still problems. I think the issue is with the routine W3GFCL, which has problems identifying the source grid cells that need to be used for interpolation to the destination grid if the longitude or latitude of the corners are the same as the destination grid point. I believe I already fixed a similar problem before, I will investigate further and find a solution.

**Comment by aliabdolali** [comment]
> After generating a new density file there are still problems. I think the issue is with the routine W3GFCL, which has problems identifying the source grid cells that need to be used for interpolation to the destination grid if the longitude or latitude of the corners are the same as the destination grid point. I believe I already fixed a similar problem before, I will investigate further and find a solution.

Any update on this?

**Comment by ukmo-juan-castillo** [comment]
I have a bug fix now for W3GFCL, but it is not the only problem (there are still some problematic points) and I need to investigate further.

**Comment by ukmo-ccbunney** [comment]
> 
> 
> I have a bug fix now for W3GFCL, but it is not the only problem (there are still some problematic points) and I need to investigate further.

Is this an existing bug that can be raised in a new issue and fixed in a separate PR? Or is it a results of the changse in this PR?
It would be good to get this staging branch in a position to be merged if not.

**Comment by ukmo-juan-castillo** [comment]
I think this is a combination of an existing bug and problems with the density.nc file as discussed before. I guess we can raise a new issue for the bug and fix the density.nc file in the present PR. I am still trying to find a final solution.

**Comment by ukmo-juan-castillo** [comment]
I created a new issue to deal with the bug that is causing part of the observed errors. The rest of the errors will be fixed by using the attached density.nc file (I simply changed the extension so that the upload is accepted):
[density.txt](https://github.com/NOAA-EMC/WW3/files/6795607/density.txt)

**Comment by aliabdolali** [comment]
I changed the density.nc and uploaded it to ftp. @ukmo-juan-castillo please update the version number in model/bin/ww3_from_ftp.sh to 7.12.5
However, I still get the same error and i think it comes from the fact that the density file does not cover the model domain completely.

**Comment by ukmo-juan-castillo** [comment]
What are the specific points in the density file that cause the problem? Please remember that you will also need the bug fix #432 for everything to work. I will double check my runs in any case.

**Comment by aliabdolali** [comment]
In that case, #432 should go before ukmet staging. @ukmo-ccbunney could you confirm? I can just check it to make sure #432 fixes this problem, or it still remains.

**Comment by ukmo-juan-castillo** [comment]
Yes please, double checking that the bug fix also solves the problems for you will be a great help. I agree regarding the order of merging.

**Comment by ukmo-ccbunney** [comment]
> 
> 
> In that case, #432 should go before ukmet staging. @ukmo-ccbunney could you confirm? I can just check it to make sure #432 fixes this problem, or it still remains.

Hopefully, the new density file has stopped the segmentation fault (due to zero density), but it is probably still giving a warning about points being outside the grid. Is that the case?

I'm happy for the bug fix to go in first.

**Comment by ukmo-juan-castillo** [comment]
Yes, there may be still genuine warnings about points outside the grid, but they should all of them now be points along the domain boundaries and dealt with properly in the code.

**Comment by aliabdolali** [comment]
for ukmet staging with the new density.nc, I can execute prnc with no problem. Howrver, looking inside the stdout of prnc_RHO.out and also others like WND, ...:



 ```
*** WAVEWATCH-III WARNING W3PRNC :
     GRID POINT     39     2  18.84  40.25
 NOT COVERED BY INPUT GRID.


 *** WAVEWATCH-III WARNING W3PRNC :
     GRID POINT     20    27  15.65  43.76
 NOT COVERED BY INPUT GRID.


 *** WAVEWATCH-III WARNING W3PRNC :
     GRID POINT      2    30  12.28  44.20
 NOT COVERED BY INPUT GRID.


 *** WAVEWATCH-III WARNING W3PRNC :
     GRID POINT     43    32  20.01  44.24
 NOT COVERED BY INPUT GRID.


 *** WAVEWATCH-III WARNING W3PRNC :
     GRID POINT     23    33  16.26  44.55
 NOT COVERED BY INPUT GRID.


 *** WAVEWATCH-III WARNING W3PRNC :
     GRID POINT     43    33  20.03  44.37
 NOT COVERED BY INPUT GRID.


 *** WAVEWATCH-III WARNING W3PRNC :
     GRID POINT     43    34  20.05  44.50
 NOT COVERED BY INPUT GRID.


 *** WAVEWATCH-III WARNING W3PRNC :
     GRID POINT     43    35  20.07  44.64
 NOT COVERED BY INPUT GRID.


 *** WAVEWATCH-III WARNING W3PRNC :
     GRID POINT     43    36  20.08  44.77
 NOT COVERED BY INPUT GRID.


 *** WAVEWATCH-III WARNING W3PRNC :
     GRID POINT     43    37  20.10  44.91
 NOT COVERED BY INPUT GRID.


 *** WAVEWATCH-III WARNING W3PRNC :
     GRID POINT     43    38  20.12  45.04
 NOT COVERED BY INPUT GRID.


 *** WAVEWATCH-III WARNING W3PRNC :
     GRID POINT     43    39  20.13  45.18
 NOT COVERED BY INPUT GRID.


 *** WAVEWATCH-III WARNING W3PRNC :
     GRID POINT     43    40  20.15  45.31
 NOT COVERED BY INPUT GRID.

```

Then I got segfault during ww3_shel execution:
```
     Updating air density

  WAVEWATCH III calculating for 2014/03/10 00:00:00 UTC at 16:33:45
forrtl: severe (174): SIGSEGV, segmentation fault occurred
Image              PC                Routine            Line        Source
ww3_shel           0000000000680BED  Unknown               Unknown  Unknown
libpthread-2.17.s  00002AAD634CE630  Unknown               Unknown  Unknown
ww3_shel           00000000004FB00E  Unknown               Unknown  Unknown
ww3_shel           00000000004F5E9A  Unknown               Unknown  Unknown
ww3_shel           00000000004A052E  Unknown               Unknown  Unknown
ww3_shel           000000000040C584  Unknown               Unknown  Unknown
ww3_shel           000000000040349E  Unknown               Unknown  Unknown
libc-2.17.so       00002AAD636FD555  __libc_start_main     Unknown  Unknown
ww3_shel           00000000004033A9  Unknown               Unknown  Unknown
```


Now, I have merged #432 to staging locally and testing ww3_tp2.15 tests with new density.nc file. I'll report soon.

**Comment by aliabdolali** [comment]
@ukmo-ccbunney @ukmo-juan-castillo 
I ran ww3_tp2.15 with combined staging and #432 and it ran successfully. So I am going to test #432 and merge it first to NOAA/develop and then we will merge staging.

**Comment by ukmo-ccbunney** [comment]
@aliabdolali That's great - thanks very much.

**Comment by aliabdolali** [comment]
> @aliabdolali That's great - thanks very much.

now testing this one.

**Comment by aliabdolali** [comment]
Hi @ukmo-juan-castillo @ukmo-ccbunney 
Here is the summary from matrix.comp on noaa hpc:
`mww3_test_03/./work_PR2_UQ_MPI_d2                     (9 files differ) `**Expected**
`mww3_test_03/./work_PR2_UNO_MPI_d2                     (7 files differ) `**Expected**
`mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (8 files differ)` **Expected**
`mww3_test_03/./work_PR3_UQ_MPI_d2                     (9 files differ) `**Expected**
`mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (10 files differ) `**Expected**
`mww3_test_03/./work_PR1_MPI_d2                     (10 files differ) `**Expected**
`mww3_test_03/./work_PR3_UNO_MPI_d2                     (10 files differ) `**Expected**
`mww3_test_07/./work_PR3_UQ                     (3 files differ) `**Expected**
`ww3_tp2.10/./work_MPI_OMPH                     (7 files differ)` **Expected**


`ww3_tnc1/./work                     (18 files differ)`


**These are related to the fix in the wind file, correct?**
```
ww3_tp2.15/./work_PR3_UQ_MPI                     (38 files differ)
ww3_tp2.15/./work_PR3_UQ_RHO_MPI                     (5 files differ)
ww3_tp2.15/./work_5km                     (5 files differ)
ww3_tp2.15/./work_PR3_UQ_RHO                     (8 files differ)
ww3_tp2.15/./work_PR3_UQ                     (6 files differ)
ww3_tp2.15/./work_MPI_5km                     (3 files differ)
```

`ww3_tp2.14/./work_OASACM6                     (1 files differ)`
**ww3_ounf.out**
 _MAPSTA output disabled for SMC grids_
`ww3_tp2.10/./work_MPI                     (1 files differ)`
**ww3_ounf.out**
 _MAPSTA output disabled for SMC grids_
`ww3_tp2.16/./work_MPI_OMPH                     (1 files differ)`
**ww3_ounf.out**
_MAPSTA output disabled for SMC grids_
`ww3_tp2.16/./work_MPI                     (1 files differ)`
**ww3_ounf.out**
 _MAPSTA output disabled for SMC grids_
`ww3_tp2.17/./work_ma1                     (1 files differ)`
**log.mww3**
_WAVEWATCH III log file                                 version 7.12_

I need to see more for these ones:
```
ww3_ufs1.1/./work_a_nth                     (16 files differ)
ww3_ufs1.1/./work_a_npl                     (16 files differ)
ww3_ufs1.1/./work_a                     (16 files differ)
ww3_ufs1.1/./work_b                     (16 files differ)
ww3_ufs1.2/./work_a                     (43 files differ)
ww3_ufs1.2/./work_b                     (43 files differ)
```

[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/6841194/matrixCompSummary.txt)
[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/6841195/matrixCompFull.txt)

**Comment by ukmo-juan-castillo** [comment]
Yes, I would expect some differences in the ww3_tp2.15 tests that depend on wind, although they should be smaller that for the test reading the air density. I build the air density from the wind file, but I corrected some points of the density file that did not contain information in wet points. The last bug fix just corrects one grid point that is not properly interpolated in these input files.

I am afraid I can not be of much help regarding the ww3_usf tests.

**Comment by aliabdolali** [comment]
> Yes, I would expect some differences in the ww3_tp2.15 tests that depend on wind, although they should be smaller that for the test reading the air density. I build the air density from the wind file, but I corrected some points of the density file that did not contain information in wet points. The last bug fix just corrects one grid point that is not properly interpolated in these input files.
> 
> I am afraid I can not be of much help regarding the ww3_usf tests.

What about ww3_tnc1, ww3_tp2.14, ww3_tp2.16, ww3_tp2.17, ww3_tp2.10?

**Comment by aliabdolali** [comment]
> Yes, I would expect some differences in the ww3_tp2.15 tests that depend on wind, although they should be smaller that for the test reading the air density. I build the air density from the wind file, but I corrected some points of the density file that did not contain information in wet points. The last bug fix just corrects one grid point that is not properly interpolated in these input files.
> 
> I am afraid I can not be of much help regarding the ww3_usf tests.

For ufs cases, I checked the results, all came from current and ice interpolation. Since the ice and current had changed, the hs and other wave results changed. it should be due to the fixes for ww3_prnc, correct?

**Comment by ukmo-ccbunney** [comment]
> > Yes, I would expect some differences in the ww3_tp2.15 tests that depend on wind, although they should be smaller that for the test reading the air density. I build the air density from the wind file, but I corrected some points of the density file that did not contain information in wet points. The last bug fix just corrects one grid point that is not properly interpolated in these input files.
> > I am afraid I can not be of much help regarding the ww3_usf tests.
> 
> What about ww3_tnc1, ww3_tp2.14, ww3_tp2.16, ww3_tp2.17, ww3_tp2.10?

@aliabdolali The differences in ww3_ounf.out are expected (it's just an extra output message)

File: model/bin/cmplr.env (hunk 1 above)
Comment (aliabdolali) [reply model/bin/cmplr.env:177]: checked on WCOSS successfully.

**Review by aliabdolali — APPROVED** [review]

**Pull request merged**